### ISSUE TITLE
[Feature request] Batch dump to PC

### ISSUE THREAD

**Issue by Corgano** [DarkMatterCore/nxdumptool#58]
Opposite of installing to PC, have it batch dump to PC. In effect, the exact opposite of nx_usbloader

**Comment by DarkMatterCore** [comment]
@Corgano Hi there. Thanks for opening this feature request issue.

It's on my to-do list, albeit with a very low priority at this moment, because of two different reasons:

1. Working on a Switch <-> PC transfer protocol will inevitably require a PC companion program capable of running under multiple OS, which means I'll have to get better at another language (Python, Java) in order to achieve that.

2. I'm busy with the ongoing nxdumptool rewrite in the private GitHub repository. Everything is really being rewritten from scratch. :P

I won't promise anything right away, but neither will I disregard this feature request. I do wish to implement this sometime in the future.

**Comment by Corgano** [comment]
Just future thinking here, but maybe when you do get around to it you could chat up the guys at nxusbloader about integrating dumping into it. They already have a payload injector and a nsp merger integrated with their pc app, and it'd be awesome to have one so that does everything. If you could figure out a protocol, you'd only have to worry about your end of things, and nxusbloader is already cross platform.

**Comment by Corgano** [comment]
I mentioned it on an issue at NS_Usbloader and he seems game to come up with something, whenever you're ready.

**Comment by DarkMatterCore** [comment]
@Corgano Just took a look at the issue and made a comment there. Hopefully we'll both come up with something nice for the nxdumptool rewrite.

**Comment by Corgano** [comment]
I must say I'm quite excited. Any idea when the rewrite will be done? Some time before christmas?

**Comment by DarkMatterCore** [comment]
No ETA at this moment, sadly. I do \*hope\* to release it before Christmas, though. Let me elaborate.

I'd say format compatibility is around 70% done, with only CNMT, NPDM, NSO and NACP support missing - luckily, these are pretty easy to implement, and shouldn't take me more than a few days.

Then again, I also have to rewrite title/content listing, which will be critical to take advantage of the new codebase - it's basically what "glues" everything else.

And don't even get me started on the new UI... I have a small PoC in a different repository with touch support, but it's far from complete - button/stick inputs are kinda wonky to handle under lvgl, and I need to optimize the current screen drawing code, because it doesn't run as smooth as I want. Basically, I have to study lvgl in-depth.

... Also, let's not forget that my Internet connection is total garbage and that I suffer from daily power outages on a regular basis. For the record, I spent more than two days trying to upload the silent update NRO. This totally hinders my daily progress.

**Comment by DarkMatterCore** [comment]
The rewrite is needed because the level of disorder in the current codebase is hideous. It was already becoming unmainteinable, and adding new features to it was a nightmare. Honestly, I just grabbed the best parts of it and decided to ditch it.

The stack corruption issue I fixed in the last release was actually what motivated me to start the rewrite - this bug was driving me crazy and I couldn't find any possible explanation for it.

Turns out I was able to reproduce the same issue in the rewrite and fixed it there before backporting the needed changes to the current codebase.

**Comment by DarkMatterCore** [comment]
Not only batch dumps will benefit from this. Everything else will do as well. 😛

**Comment by DarkMatterCore** [comment]
@Corgano Status update: Switch <-> PC communication now works in the rewrite codebase, thanks to a debug tool written in Java by developersu. 🙂

Actual tests haven't been carried out because the companion app still needs to be developed, but the groundwork for this feature has mostly been finished (Switch side of things, at least).

Combined with the new threaded I/O, it's gonna be interesting to see how fast will USB dumping perform. My full Super Mario Odyssey gamecard is dumped in ~ 4 minutes with 20 seconds using the SD card.

**Comment by DarkMatterCore** [comment]
@Corgano USB I/O is pretty much complete at this point.

![ns-usbloader](https://cdn.discordapp.com/attachments/478162551381819392/708996413123723315/Screenshot_9.png)

![nxdumptool](https://cdn.discordapp.com/attachments/478162551381819392/708998630207651881/2020051007043500-1E1800B8D04F999C436DDFE2B8CD0B81.jpg)

Extracted Dark Souls RomFS dump via USB. Took 97 secs.

Other tests have been carried out as well, like dumping my SMO gamecard (took 3:20 minutes) and SSBU's full RomFS image (took 5:45 minutes).

This issue will be kept open until the rewrite is finished.

**Comment by Corgano** [comment]
That is fantastic! As soon as you have something stable enough to test I'd be happy to test it.

**Comment by Corgano** [comment]
Will this feature need to wait for the rewrite to be completed, or is it possible to back this to the current release version in the meantime?

**Comment by DarkMatterCore** [comment]
@Corgano It will only be available in the rewrite. It could be possible to port this back to the old nxdumptool branch, but honestly I don't want to lose any more time glueing new code to it.

**Comment by Corgano** [comment]
Any news on the rewrite? Which discord do you usually hang out at?

**Comment by DarkMatterCore** [comment]
@Corgano Still working on it at irregular intervals, sadly. Power outages in my city aren't very helpful. Feel free to send me a DM or a friend request on Discord, they're both open to anyone. It's easier than trying to contact me through a server.

**Comment by Corgano** [comment]
I don't know how to friend people directly on discord.

Just curious is it faster dumping over usb or to SD?

**Comment by DarkMatterCore** [comment]
@Corgano Orders of magnitude faster. @averne tested ACNH's RomFS dumping via USB and it took less than 6 minutes to dump all 60k+ files.

**Comment by Corgano** [comment]
That sounds impressive, how long would that take to dump to SD?

I'd be happy to test dumping stuff once you have a test build available

**Comment by DarkMatterCore** [comment]
@Corgano More than 2 hours for most people. This is because random access writes are way slower for SD cards, thus dumping many small files takes a lot of time. We're talking about 60k+ files, and most of them don't even reach 500 KiB in size.

**Comment by Corgano** [comment]
Holy crap that's insane, like a 2,000% boost.

**Comment by DarkMatterCore** [comment]
@Corgano Indeed. Speed improvements for sequential access writes are not as big, but they're still very much noticeable.

**Comment by DarkMatterCore** [comment]
@Corgano Some days ago, I decided to make the rewrite codebase public as a separate branch in this repository. If you ever feel like testing this feature yourself, by all means go ahead. If you hit me up at Discord, I'll happily assist you.

**Comment by PolyCatDev** [comment]
Imma jump in here and show off a client for Linux. I've been using it and it works wonders: https://flathub.org/apps/org.v1993.NXDumpClient

**Comment by halotroop2288** [comment]
This is complete now, right?

Is there a client for this on Windows yet?

**Comment by DarkMatterCore** [comment]
> This is complete now, right?
> 
> Is there a client for this on Windows yet?

Yes, but you need to use a [rewrite branch build](https://github.com/DarkMatterCore/nxdumptool/releases/tag/rewrite-prerelease).

We offer a [Python script](https://github.com/DarkMatterCore/nxdumptool/tree/rewrite/host) that can be used under Windows. It requires a USB driver installation with Zadig.

We also offer EXE builds of this script in our Discord server.

**Comment by MAProsper** [comment]
Would it be plausible, instead of requiring a Python app, to this via a web browser?

Similar to how it's already possible to launch fusée via WebUSB,
one could connect to the switch with WebUSB and trigger a web download stream.

**Comment by DarkMatterCore** [comment]
> Would it be plausible, instead of requiring a Python app, to this via a web browser?
> 
> Similar to how it's already possible to launch fusée via WebUSB, one could connect to the switch with WebUSB and trigger a web download stream.

I'm not too well-versed in modern web development, sadly. I'll say, though, that using regular web downloads as the transfer mechanism could be an issue, especially when dealing with NSP transfers.

The USB ABI itself is [open and well-documented](https://github.com/DarkMatterCore/nxdumptool/blob/rewrite/host/README.md). I encourage people to come up with their own tools to interact with nxdumptool.

**Comment by Whovian9369** [comment]
Switching exclusively to a WebUSB implementation would also cut off anyone that does not use Chrome-based browsers,which would include myself.

**Comment by MAProsper** [comment]
True, switching exclusively to a web-based solution could be a bad idea.
Plus having a headless option, and a more flexible dev environment with Python is nice.
I just thought it would be a more convenient end-user option.

Having the USB ABI clearly documented I might give it a shot in a few months,
it sound like a fun project and a opportunity to learn a bunch of new stuff.

**Comment by DarkMatterCore** [comment]
> True, switching exclusively to a web-based solution could be a bad idea. Plus having a headless option, and a more flexible dev environment with Python is nice. I just thought it would be a more convenient end-user option.
> 
> Having the USB ABI clearly documented I might give it a shot in a few months, it sound like a fun project and a opportunity to learn a bunch of new stuff.

Be my guest. Feel free to join our Discord server as well -- I did my best to document the USB ABI as good as possible, but there might be some implementation details I unintentionally skipped (I hope not).

Discussing these matters in realtime is just way better.

**Comment by MAProsper** [comment]
Hi again, just finished creating the web-based implementation of the USB host tool!

It has full feature parity with the original, plus some minor nice-to-haves, and it works on any OS with a Chromium-based browser.

>  I did my best to document the USB ABI as good as possible, but there might be some implementation details I unintentionally skipped (I hope not).

It was very well documented, plus having the original Python script helped alot in the porting!

Would it be okay if I linked it here or in the Discord server?

**Comment by DarkMatterCore** [comment]
> Hi again, just finished creating the web-based implementation of the USB host tool!
> 
> It has full feature parity with the original, plus some minor nice-to-haves, and it works on any OS with a Chromium-based browser.
> 
> Would it be okay if I linked it here or in the Discord server?

Sure, be my guest. Linking it here or sharing it in the Discord server are both fine by me.

I'll test it as soon as I get back home.

**Comment by MAProsper** [comment]
Feedback or testing is well appreciated!
I'm pretty busy with work but I will fix what I can.

Web-App: https://nxdt.m16r.xyz/
Repository: https://github.com/MAProsper/web-nxdt
